tools/payloads/configbytes/apk_configbytes_extract.rb

```diff
@@ -0,0 +1,231 @@
+  bytes[0]
+
+  File.open(text_file, 'wb') do |f|
+    f.write(bytes[0])
+  end
+
+  $stderr.puts "[+] Saved as: #{text_file}"
+end
+
+def check_tools(verbose)
+  # CHECK IF ALL TOOLS ARE INCLUDED BEFORE STARTING
+  tools = ["unzip", "d2j-dex2jar", "java", "apktool"]
+
+  tools.each do |tool|
+    path = Rex::FileUtils.find_full_path(tool)
+    if path && ::File.file?(path)
+      $stderr.puts "[+] Tool present: #{path}" if verbose
+    else
+      raise RuntimeError, "[-] #{tool} command not found."
+    end
+  end
+end
+
+def parse_args(args)
+  opts = {}
+  opt = OptionParser.new
+  banner = "apk_configbytes_extract - a tool to extract configbytes from an android payload.\n"
+  banner << "Usage: #{$0} < -a apk-file -o path -j path > [options]\n"
+  banner << "Example: #{$0} -a metasploit.apk -o /root/configbytes.txt -j /root/fernflower.jar"
+  opt.banner = banner
+  opt.separator('')
+  opt.separator('Options:')
+
+  opt.on('-a', '--apk             <path>', String, 'Specify apk to extract configbytes') do |a|
+    opts[:apk] = a
+  end
+
+  opt.on('-o', '--out             <path>', String, 'Save configbytes to a file') do |o|
+    opts[:out] = o
+  end
+
+  opt.on('-j', '--jar             <path>', String, 'Specify fernflower path') do |j|
+    opts[:jar] = j
+  end
+
+  opt.on('-v', '--verbose', 'Displays verbose output') do
+    opts[:verbose] = true
+  end
+
+  opt.on('-k', '--keep', 'Keep working directory') do
+    opts[:keep] = true
+  end
+
+  opt.on_tail('-h', '--help', 'Show this message') do
+    raise HelpError, "#{opt}"
+  end
+
+  begin
+    opt.parse!(args)
+  rescue OptionParser::InvalidOption => e
+    raise UsageError, "Invalid option\n#{opt}"
+  rescue OptionParser::MissingArgument => e
+    raise UsageError, "Missing required argument for option\n#{opt}"
+  end
+
+  if opts.empty?
+    raise UsageError, "No options\n#{opt}"
+  end
+
+  if opts[:apk].nil?
+    raise UsageError, "Missing required argument apk file\n#{opt}"
+    exit(1)
+  end
+
+  if opts[:jar].nil?
+    raise UsageError, "Missing required argument fernflower path\n#{opt}"
+    exit(1)
+  end
+
+  opts
+end
+
+begin
+  options = parse_args(ARGV)
+rescue HelpError => e
+  $stderr.puts e.message
+  exit(1)
+rescue ExtractError => e
+  $stderr.puts "Error: #{e.message}"
+  exit(1)
+end
+
+begin
+  verbose = options[:verbose]
+
+  check_tools(verbose)
+
+  temp_dir = Dir.mktmpdir
+  temp_dir << "/"
+
+  if options[:out].nil?
+    $stderr.puts "[-] No output option selected, working in #{temp_dir}"    
+    output = "#{temp_dir}configbytes.txt"
+    options[:keep] = true
+  else
+    output = options[:out]
+  end
+
+  apk = options[:apk]
+  zip_file = "#{temp_dir}temp.zip"
+  temp_apk = "#{temp_dir}temp.apk"
+  FileUtils.cp(apk, temp_apk)
+
+  $stderr.puts "[+] Renaming apk file to zip file"
+  File.rename(temp_apk, zip_file)
+
+  $stderr.puts "[+] Using unzip on zip file for dex file"
+  unzip_cmd = run_cmd("unzip", "#{temp_dir}temp.zip -d #{temp_dir}", verbose)
```

Comment:
A note for future travelers, the `unzip` command since 5.50 strips "parent dir" path components meaning this should not be vulnerable to path traversal attacks.

<details>
<summary>See this MAN page excerpt from unzip v6.00 on Kali Linux</summary>

```
       -:     [all but Acorn, VM/CMS, MVS, Tandem] allows to extract archive members into locations outside of the current `` extraction root folder''. For security reasons, unzip normally removes ``parent dir'' path components (``../'') from the names of extracted file.  This safety feature (new for version 5.50)  prevents  unzip  from  accidentally
              writing files to ``sensitive'' areas outside the active extraction folder tree head.  The -: option lets unzip switch back to its previous, more liberal behaviour, to allow exact extraction of (older) archives that used ``../'' components to create multiple directory trees at the level of the current extraction folder.  This option does
              not enable writing explicitly to the root directory (``/'').  To achieve this, it is necessary to set the extraction target folder to root (e.g. -d / ).  However, when the -: option is specified, it is still possible to implicitly write to the root directory by specifying enough ``../'' path components within the zip archive.  Use  this
              option with extreme caution.
```
</details>